File: chat/src/lib/calls/tie-break.ts
const textEncoder = new TextEncoder();

export function compareOctetStrings(left: string, right: string): number {
  const leftBytes = textEncoder.encode(left);
  const rightBytes = textEncoder.encode(right);
  const len = Math.min(leftBytes.length, rightBytes.length);
  for (let i = 0; i < len; i += 1) {
    const diff = leftBytes[i] - rightBytes[i];
    if (diff !== 0) return diff;
  }
  return leftBytes.length - rightBytes.length;
}

function bareCallJid(jid: string): string {
  return (jid.split("/")[0] ?? "").toLowerCase();
}

export function isSameCallBareJid(left: string, right: string): boolean {
  return bareCallJid(left) === bareCallJid(right);
}

export function incomingProposeWinsTieBreak(
  incomingSid: string,
  outgoingSid: string,
  incomingFrom: string,
  ourFullJid: string,
): boolean {
  const sidOrder = compareOctetStrings(incomingSid, outgoingSid);
  if (sidOrder !== 0) return sidOrder < 0;
  return compareOctetStrings(incomingFrom, ourFullJid) < 0;
}

File: apps/android/core/client/src/main/kotlin/social/waddle/android/client/calls/TieBreak.kt
package social.waddle.android.client.calls

/**
 * XEP-0353 tie-breaking primitives, a byte-exact port of the web
 * client's tie-break.ts. §Tie Breaking (anchor tie-break-1): the lower
 * of the two session ids wins, where "lower" means sorted first under
 * RFC 4790 §9.3 "i;octet" collation — an unsigned byte-wise compare of
 * the UTF-8 encodings, shorter string first on a shared prefix.
 */
internal fun compareOctetStrings(left: String, right: String): Int {
    val leftBytes = left.encodeToByteArray()
    val rightBytes = right.encodeToByteArray()
    val len = minOf(leftBytes.size, rightBytes.size)
    for (i in 0 until len) {
        val diff = (leftBytes[i].toInt() and BYTE_MASK) - (rightBytes[i].toInt() and BYTE_MASK)
        if (diff != 0) return diff
    }
    return leftBytes.size - rightBytes.size
}

private fun bareCallJid(jid: String): String = jid.substringBefore('/').lowercase()

/** Case-insensitive bare-JID equality for tie-break peer matching. */
internal fun isSameCallBareJid(left: String, right: String): Boolean =
    bareCallJid(left) == bareCallJid(right)

/**
 * Whether the INCOMING propose overrules our outgoing one: true when
 * the incoming sid is lower (XEP-0353 tie-break-1), falling back to
 * the JID comparison for the unlikely equal-sid case.
 */
internal fun incomingProposeWinsTieBreak(
    incomingSid: String,
    outgoingSid: String,
    incomingFrom: String,
    ourFullJid: String,
): Boolean {
    val sidOrder = compareOctetStrings(incomingSid, outgoingSid)
    if (sidOrder != 0) return sidOrder < 0
    return compareOctetStrings(incomingFrom, ourFullJid) < 0
}

private const val BYTE_MASK = 0xFF
